README
BankAccount-Validator-UK

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After  installing, you  can  find documentation for this module with  the perldoc
command.

    perldoc BankAccount::Validator::UK

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=BankAccount-Validator-UK

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/BankAccount-Validator-UK

    CPAN Ratings
        http://cpanratings.perl.org/d/BankAccount-Validator-UK

    Search CPAN
        http://search.cpan.org/dist/BankAccount-Validator-UK/

LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2014 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for BankAccount::Validator::UK

0.54  2020-06-24T13:50:00:+01:00
      - Updated the valacdos.txt (v6.12)

0.53  2020-04-07T19:05:00:+01:00
      - Updated the valacdos.txt (v6.10)

0.52  2020-03-04T16:10:00:+01:00
      - Updated the valacdos.txt (v6.00)

0.51  2020-01-15T14:15:00:+01:00
      - Updated the valacdos.txt (v5.90)

0.50  2019-09-08T07:30:00:+01:00
      - Moved author test scripts to xt/ subfolder.

0.49  2019-09-05T14:35:00:+01:00
      - Updated the valacdos.txt (v5.80)

0.48  2019-05-19T03:45:00:+01:00
      - Fixed the source data valacdos.txt.

0.47  2019-04-30T11:30:00:+01:00
      - Updated the valacdos.txt (v5.70)

0.46  2019-02-15T12:35:00:+01:00
      - Updated the valacdos.txt (v5.50)

0.45  2018-12-21T12:00:00:+01:00
      - Updated the valacdos.txt (v5.40)

0.44  2018-11-05T17:30:00:+01:00
      - Updated the valacdos.txt (v5.30)

0.43  2018-09-21T11:20:00:+01:00
      - Updated the valacdos.txt (v5.20)

0.42  2018-08-20T11:30:00:+01:00
      - Updated the valacdos.txt (v5.10)

0.41  2018-07-16T19:00:00:+01:00
      - Used File::ShareDir v1.114 instead of File::Share.
      - Explicitly added configure prereq File::ShareDir::Install v0.13.

0.40  2018-07-15T12:10:00:+01:00
      - Updated the valacdos.txt (v5.00)

0.39  2018-05-22T11:55:00:+01:00
      - Updated the valacdos.txt (v4.90)

0.38  2018-03-16T18:11:00:+01:00
      - Updated pod with regard to the changes to valacdos.txt v4.80.

0.37  2018-03-16T18:10:00:+01:00
      - Updated the valacdos.txt (v4.80)
      - Added sort code range 232290-232290 for Mod11.
      - Added sort code range 040082-040082 for Mod10.

0.36  2018-02-26T10:30:00:+01:00
      - Improved data loading.

0.35  2018-02-25T16:30:00:+01:00
      - Used namespace::autoclean instead.

0.34  2018-02-23T15:35:00:+01:00
      - Updated the valacdos.txt (v4.70)
      - Added sort code range 302880 - 302880 for Mod11.

0.33  2017-09-14T11:25:00:+01:00
      - Updated the valacdos.txt (v4.60)
      - Added sort code range 040072-040073
      - Changed sort code range 040406-040406 to 040400-041399.
      - Amended sort code range 040010-040059 to 040020-040059.
      - Amended sort code range 040010-040059 to 040010-040014 for both Mod11 and DblAl.
      - Deleted 040015-040019.
      - Added sort code range 236972-236972.
      - Added sort codes - 231228, 231618 and 233135 for both Mod11 and DblAl.

0.32  2017-09-03T12:00:00:+01:00
      - Updated copyright year information.

0.31  2017-07-27T10:30:00:+01:00
      - Added new sort codes 230088 and 040004. (v4.40)

0.30  2017-06-10T13:50:00:+01:00
      - Updated the valacdos.txt dated 2017-07-03.
      - Removed the following sorting codes from the range
        400516 - 404799 DBIAI: 401055, 401199, 401266, 401276 - 401279, 401900, 401950 and 404375 - 404384.
      - Amended the weight for the following sorting codes from the range
        400516 - 404788 Mod 11: 401055, 401199, 401266, 401276 - 401279, 401900, 401950 and 404375 - 404384.

0.29  2017-05-11T19:55:00:+01:00
      - Addition of new sorting codes. (v4.20)
      - Updated the valacdos.txt dated 12th Jun'2017.
      - Added support for Clear Bank and HSBC.

0.28  2017-01-28T20:35:00:+01:00
      - Addition of new sorting codes. (v4.10)
      - Updated the valacdos.txt dated 6th Mar'2017.
      - Added support for Charity Bank, Contis Financial Services, Starling Bank and Tesco Bank.
      - Removed support for HSBC.

0.27  2016-12-04T19:00:00:+01:00
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.26  2016-11-29T12:20:00:+01:00
      - Addition of new sorting codes. (v4.00)
      - Updated the valacdos.txt dated 9th Jan'2017.

0.25  2016-06-04T11:45:00:+01:00
      - Addition of new sorting codes. (v3.90)
      - Removal of sorting codes 400194 - 400195 (v3.90)
      - Tidied up unit test scripts.

0.24  2016-02-05T14:20:00:+01:00
      - Updated the valacdos.txt dated 15th Feb'2016.
      - Updated Copyright year information in the pod.

0.23  2015-11-28T02:50:00:+01:00
      - Added key 'provides' to the Makefile.PL script.

0.22  2015-11-09T12:35:00:+01:00
      - Updated the valacdos.txt dated 7th Dec'2015.

0.21  2015-10-12T11:20:00:+01:00
      - Updated the valacdos.txt dated 26th Oct'2015.

0.20  2015-08-13T10:20:00:+01:00
      - Updated link to the VOCALINK site as well the link to the PDF document
      describing the "Modulus Checking".

0.19  2015-07-29T12:00:00:+01:00
      - Updated the valacdos.txt dated 20th Jul'2015.
      - Added support for the following banks:
        - Hoares Bank
        - HSBC
        - Metro Bank
        - Unity Trust Bank
        - Yorkshire Bank

0.18  2015-03-19T16:48:00:+01:00
      - Added LICENSE file.

0.17  2015-03-01T08:30:00:+01:00
      - Added File::ShareDir::Install as configure requires in the Makefile.PL file.
      - Updated description of method is_valid() in the pod document.

0.16  2015-02-24T17:20:00:+01:00
      - Tidied up pod document.

0.15  2015-02-24T10:15:00:+01:00
      - Updated document v3.20 valid from 23rd Mar'2015.
      - Added support for "Orwell Union Ltd.".
      - Splitted "Lloyds TSB" into two separate entities "Llyods" and "TSB".
      - Removed support for "Hoares Bank", "HSBC" and "Yorkshire Bank".

0.14  2015-01-16T14:50:15+01:00
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

0.13  2014-10-28T13:30:25+01:00
      - Updated Changes file according to CPAN::Changes::Spec package.
      - Added unit test changes.t file.

0.12  2014-10-23T23:48:10+01:00
      - Made the package File::ShareDir::Install pre-requisite in the Makefile.PL file.

0.11  2014-10-19T09:07:08+01:00
      - Minor cosmetic changes to pod document.

0.10  2014-10-18T15:50:10+01:00
      - Updated valacdos.txt file dated 6th Oct'2014.

0.09  2014-10-18T15:27:39+01:00
      - Extracted the rule to separate files i.e. valacdos.txt and scsubtab.txt.

lib/BankAccount/Validator/UK.pm
package BankAccount::Validator::UK;

$BankAccount::Validator::UK::VERSION   = '0.54';
$BankAccount::Validator::UK::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

BankAccount::Validator::UK - Interface to validate UK bank account.

=head1 VERSION

Version 0.54

=cut

use 5.006;
use Data::Dumper;
use BankAccount::Validator::UK::Rule;

use Moo;
use namespace::autoclean;

has sc         => (is => 'rw');
has an         => (is => 'rw');
has mod        => (is => 'rw');
has attempt    => (is => 'rw');
has last_ex    => (is => 'rw');
has trace      => (is => 'rw');
has debug      => (is => 'ro', default => sub { 0 });
has last_check => (is => 'rw', default => sub { 0 });
has multi_rule => (is => 'ro', default => sub { 0 });
has sort_code  => (is => 'ro', default => sub { BankAccount::Validator::UK::Rule::get_sort_codes() });

=head1 DESCRIPTION

The module uses the algorithm provided by VOCALINK to validate the bank sort code
and account number.  It is  done by modulus  checking  method as specified in the
document which is available on their website L<VOCALINK|https://www.vocalink.com/customer-support/modulus-checking>
It currently supports the document L<v6.12|https://www.vocalink.com/media/4475/vocalink-validating-account-numbers-v612.pdf> drafted 12th Jun 2020.

Institutions covered by this document are below:

=over 4

=item * Allied Irish

=item * Bank of England

=item * Bank of Ireland

=item * Bank of Scotland

=item * Barclays

=item * Bradford and Bingley Building Society

=item * Charity Bank

=item * Citibank

=item * Clear Bank

=item * Clydesdale

=item * Contis Financial Services

=item * Co-Operative Bank

=item * Coutts

=item * First Trust

=item * Halifax

=item * Hoares Bank

=item * HSBC

=item * Lloyds

=item * Metro Bank

=item * NatWest

=item * Nationwide Building Society

=item * Northern

=item * Orwell Union Ltd.

=item * Royal Bank of Scotland

=item * Santander

=item * Secure Trust

=item * Starling Bank

=item * Tesco Bank

=item * TSB

=item * Ulster Bank

=item * Unity Trust Bank

=item * Virgin Bank

=item * Williams & Glyn

=item * Woolwich

=item * Yorkshire Bank

=back

=head2 NOTE

If the modulus check shows the account number as valid this means that the account
number  is a possible account number for the sorting code but does'nt necessarily
mean that it's an account number  being  used at that sorting code.  Any  account
details found as invalid should be checked with the account holder where possible.

=head1 CONSTRUCTOR

The constructor  simply expects debug flag, which is optional. By the default the
debug flag is off.

    use strict; use warnings;
    use BankAccount::Validator::UK;

    # Debug is turned off.
    my $account1 = BankAccount::Validator::UK->new;

    # Debug is turned on.
    my $account2 = BankAccount::Validator::UK->new(debug => 1);

=head1 METHODS

=head2 is_valid($sort_code, $account_number)

It expects two parameters i.e. the sort code and the account number.The sort code
can be either nn-nn-nn or nnnnnn format. If the account number starts with 0 then
its advisable to pass in as string i.e. '0nnnnnnn'.

    use strict; use warnings;
    use BankAccount::Validator::UK;

    my $account = BankAccount::Validator::UK->new;
    print "[10-79-99][88837491] is valid.\n"
        if $account->is_valid(107999, 88837491);

    print "[18-00-02][00000190] is valid.\n"
        if $account->is_valid('18-00-02', '00000190');

=cut

sub is_valid {
    my ($self, $sc, $an) = @_;

    die("ERROR: Missing bank sort code.\n")      unless defined $sc;
    die("ERROR: Missing bank account number.\n") unless defined $an;

    ($sc, $an) = _prepare($sc, $an);
    die("ERROR: Invalid sort code.\n")      unless (length($sc) == 6);
    die("ERROR: Invalid account number.\n") unless (length($an) == 8);

    my $_sort_code = _init('u', $sc);
    my $_account_number = _init('a', $an);
    my $_rules = _get_rules($sc);

    next if (scalar(@{$_rules}) == 0);

    $self->{sc} = $sc;
    $self->{an} = $an;
    $self->{multi_rule} = (scalar(@{$_rules}) > 1)?(1):(0);
    foreach my $_rule (@{$_rules}) {
        $self->{attempt}++;
        _init('u', '090126', $_sort_code)
            if ($_rule->{ex} == 8);

        if (($_rule->{ex} == 6)
            &&
            ($_account_number->{a} =~ /^[4|5|6|7|8]$/)
            &&
            ($_account_number->{g} == $_account_number->{h})) {

            $self->{last_ex} = $_rule->{ex};
            $self->{last_check} = 1;
            push @{$self->{trace}}, {'ex'  => $_rule->{ex},
                                     'mod' => $_rule->{mod},
                                     'res' => 'VALID'};
            next;
        }

        if (($_rule->{ex} == 7) && ($_account_number->{g} == 9)) {
            _init('u','000000', $_rule);
            _init('a','00', $_rule);
        }
        elsif ($_rule->{ex} == 8) {
            _init('u', '090126', $_sort_code);
        }
        elsif ($_rule->{ex} =~ /^[2|9]$/) {
            if ($_rule->{ex} == 9) {
                _init('u', '309634', $_sort_code);
            }
            elsif ($_account_number->{a} != 0) {
                if ($_account_number->{g} != 9) {
                    _init('u','001253', $_rule);
                    _init('a','6,4,8,7,10,9,3,1', $_rule);
                }
                elsif ($_account_number->{g} == 9) {
                    _init('u','000000', $_rule);
                    _init('a','0,0,8,7,10,9,3,1', $_rule);
                }
            }
        }
        elsif ($_rule->{ex} == 10) {
            my $_ab = sprintf("%s%s", $_account_number->{a}, $_account_number->{b});
            if ((($_ab eq "09") or ($_ab eq "99")) && ($_account_number->{g} == 9)) {
                _init('u', '000000', $_rule);
                _init('a', '00', $_rule);
            }
        }
        elsif ($_rule->{ex} == 3) {
            $self->{last_ex} = 3;
            next if ($_account_number->{c} =~ /^[6|9]$/);
        }
        elsif ($_rule->{ex} == 5) {
            _init('u', $self->{sort_code}->{$sc}, $_sort_code)
                if (exists $self->{sort_code}->{$sc});
        }

        my $_status;
        if ($_rule->{mod} =~ /MOD(\d+)/i) {
            $_status = $self->_standard_check($_sort_code, $_account_number, $_rule);
        }
        elsif ($_rule->{mod} =~ /DBLAL/i) {
            $_status = $self->_double_alternate_check($_sort_code, $_account_number, $_rule);
        }

        if (defined $_status) {
            $self->{last_ex} = $_status->{ex};
            $self->{last_check} = ($_status->{res} eq 'PASS')?(1):(0);;
            push @{$self->{trace}}, $_status;
        }

        my $_result = $self->_check_result();
        return $_result if defined $_result;
    }

    return $self->{last_check}
        if ((defined $self->{last_ex}) && ($self->{last_ex} =~ /^6$/) && ($self->{multi_rule}));

    return;
}

=head2 get_trace()

Returns the trace information about each rule that applied to the given sort code
and account number.

    use strict; use warnings;
    use Data::Dumper;
    use BankAccount::Validator::UK;

    my $account = BankAccount::Validator::UK->new;
    print "[87-14-27][09123496] is valid.\n"
        if $account->is_valid('871427', '09123496');

    print "Trace information:\n" . Dumper($account->get_trace);

=cut

sub get_trace {
    my ($self) = @_;

    return $self->{trace} if scalar(@{$self->{trace}});
}

#
#
# PRIVATE METHODS

sub _standard_check {
    my ($self, $_sort_code, $_account_number, $_rule) = @_;

    my $total = 0;
    $total += 27 if ($_rule->{ex} == 1);

    if ($_rule->{mod} =~ /MOD(\d+)/i) {
        foreach (keys %{$_sort_code}) {
            print "KEY: [$_] SC: [$_sort_code->{$_}] WEIGHTING: [$_rule->{$_}]\n"
                if $self->{debug};
            $total += $_sort_code->{$_} * $_rule->{$_};
        }

        foreach (keys %{$_account_number}) {
            print "KEY: [$_] AN: [$_account_number->{$_}] WEIGHTING: [$_rule->{$_}]\n"
                if $self->{debug};
            $total += $_account_number->{$_} * $_rule->{$_};
        }

        my $remainder = $total % $1;
        if ($_rule->{ex} == 4) {
            my $_gh = sprintf("%d%d", $_account_number->{g}, $_account_number->{h});
            if ($remainder == $_gh) {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'PASS'};
            }
        }
        elsif (($_rule->{ex} == 5) && ($1 == 11)) {
            if ($remainder == 0) {
                if ($_account_number->{g} == 0) {
                    return {'ex'  => $_rule->{ex},
                            'mod' => $_rule->{mod},
                            'rem' => $remainder,
                            'tot' => $total,
                            'res' => 'PASS'};
                }
                else {
                    return {'ex'  => $_rule->{ex},
                            'mod' => $_rule->{mod},
                            'rem' => $remainder,
                            'tot' => $total,
                            'res' => 'FAIL'};
                }
            }
            elsif ($remainder == 1) {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'FAIL'};
            }
            else {
                $remainder = 11 - $remainder;
                if ($_account_number->{g} == $remainder) {
                    return {'ex'  => $_rule->{ex},
                            'mod' => $_rule->{mod},
                            'rem' => $remainder,
                            'tot' => $total,
                            'res' => 'PASS'};
                }
                else {
                    return {'ex'  => $_rule->{ex},
                            'mod' => $_rule->{mod},
                            'rem' => $remainder,
                            'tot' => $total,
                            'res' => 'FAIL'};
                }
            }
        }
        elsif ($remainder == 0) {
            return {'ex'  => $_rule->{ex},
                    'mod' => $_rule->{mod},
                    'rem' => $remainder,
                    'tot' => $total,
                    'res' => 'PASS'};
        }
        else {
            if ($_rule->{ex} == 14) {
                if ($_account_number->{h} =~ /^[0|1|9]$/) {
                    my $an = substr($self->{an}, 0, 7);
                    $an = sprintf("%s%s", '0', $an);
                    _init('a', $an, $_account_number);

                    $total = 0;
                    foreach (keys %{$_sort_code}) {
                        print "KEY: [$_] SC: [$_sort_code->{$_}] WEIGHTING: [$_rule->{$_}]\n"
                            if $self->{debug};
                        $total += $_sort_code->{$_} * $_rule->{$_};
                    }

                    foreach (keys %{$_account_number}) {
                        print "KEY: [$_] AN: [$_account_number->{$_}] WEIGHTING: [$_rule->{$_}]\n"
                            if $self->{debug};
                        $total += $_account_number->{$_} * $_rule->{$_};
                    }

                    $remainder = $total % 11;
                    if ($remainder == 0) {
                        return {'ex'  => $_rule->{ex},
                                'mod' => $_rule->{mod},
                                'rem' => $remainder,
                                'tot' => $total,
                                'res' => 'PASS'};
                    }
                    else {
                        return {'ex'  => $_rule->{ex},
                                'mod' => $_rule->{mod},
                                'rem' => $remainder,
                                'tot' => $total,
                                'res' => 'FAIL'};
                    }
                }
                else {
                    return {'ex'  => $_rule->{ex},
                            'mod' => $_rule->{mod},
                            'rem' => $remainder,
                            'tot' => $total,
                            'res' => 'FAIL'};
                }
            }
            else {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'FAIL'};
            }
        }
    }

    return;
}

sub _double_alternate_check {
    my ($self, $_sort_code, $_account_number, $_rule) = @_;

    my $total = 0;
    $total += 27 if ($_rule->{ex} == 1);

    foreach (keys %{$_sort_code}) {
        $total += _dbal_total($_sort_code->{$_} * $_rule->{$_});
    }

    foreach (keys %{$_account_number}) {
        $total += _dbal_total($_account_number->{$_} * $_rule->{$_});
    }

    my $remainder = $total % 10;
    if ($_rule->{ex} == 1) {
        if ($remainder == 0) {
            return {'ex'  => $_rule->{ex},
                    'mod' => $_rule->{mod},
                    'rem' => $remainder,
                    'tot' => $total,
                    'res' => 'PASS'};
        }
        else {
            return {'ex'  => $_rule->{ex},
                    'mod' => $_rule->{mod},
                    'rem' => $remainder,
                    'tot' => $total,
                    'res' => 'FAIL'};
        }
    }
    elsif ($_rule->{ex} == 5) {
        if ($remainder == 0) {
            if ($_account_number->{h} == 0) {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'PASS'};
            }
        }
        else {
            $remainder = 10 - $remainder;
            if ($_account_number->{h} == $remainder) {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'PASS'};
            }
            else {
                return {'ex'  => $_rule->{ex},
                        'mod' => $_rule->{mod},
                        'rem' => $remainder,
                        'tot' => $total,
                        'res' => 'FAIL'};
            }
        }
    }
    elsif ($remainder == 0) {
        return {'ex'  => $_rule->{ex},
                'mod' => $_rule->{mod},
                'rem' => $remainder,
                'tot' => $total,
                'res' => 'PASS'};
    }
    else {
        return {'ex'  => $_rule->{ex},
                'mod' => $_rule->{mod},
                'rem' => $remainder,
                'tot' => $total,
                'res' => 'FAIL'};
    }
}

sub _init {
    my ($index, $data, $init) = @_;

    if ($data =~ /\,/) {
        map { $init->{$index++} = $_; } split /\,/,$data;
    }
    else {
        map { $init->{$index++} = $_; } split //,$data;
    }

    return $init;
}

sub _check_result {
    my ($self) = @_;

    if ($self->{multi_rule}) {
        if (((defined $self->{last_ex})
             && ($self->{last_ex} =~ /^2|10|12$/)
             && ($self->{last_check} == 1))
            ||
            ((defined $self->{last_ex})
             && ($self->{last_ex} =~ /^9|11|13$/)
             && ($self->{last_check} == 1)
             && ($self->{attempt} == 2))) {
            return 1;
        }
        elsif ((defined $self->{last_ex})
               && ($self->{last_ex} =~ /^5|6$/)
               && ($self->{last_check} == 0)) {
            return 0;
        }
        elsif ((defined $self->{last_ex})
               && ($self->{last_ex} == 0)
               && ($self->{last_check} == 1)) {
            return 1;
        }
        elsif ($self->{attempt} == 2) {
            return $self->{last_check};
        }
    }
    else {
        return $self->{last_check};
    }

    return;
}

sub _get_rules {
    my ($sc) = @_;

    return unless (defined($sc) && ($sc =~ /^\d+$/));

    my $rules;
    foreach (@{BankAccount::Validator::UK::Rule::get_rules()}) {
        push @{$rules}, $_ if ($sc >= $_->{start} && $sc <= $_->{end});
    }

    return $rules;
}

sub _dbal_total {
    my ($_total) = @_;

    if ($_total > 9) {
        my ($left, $right) = split //, $_total;
        return ($left + $right);
    }
    else {
        return $_total;
    }
}

sub _prepare {
    my ($sc, $an) = @_;

    $sc =~ s/[\-\s]+//g;
    $an =~ s/\s+//g;

    die("ERROR: Invalid bank sort code [$sc].\n")      unless ($sc =~ /^\d+$/);
    die("ERROR: Invalid bank account number [$an].\n") unless ($an =~ /^\d+$/);

    if (length($an) == 10) {
        if ($an =~ /^(\d+)\-(\d+)/) {
            $an = $2;
        }
        else {
            $an = substr($an, 0, 8);
        }
    }
    elsif (length($an) == 9) {
        my $_a = substr($an, 0, 1);
        $an = substr($an, 1, 8);
        $sc = substr($sc, 0, 5);
        $sc .= $_a;
    }
    elsif (length($an) == 7) {
        $an = '0'.$an;
    }
    elsif (length($an) == 6) {
        $an = '00'.$an;
    }

    return ($sc, $an);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/BankAccount-Validator-UK>

=head1 BUGS

Please  report  any bugs or feature requests to C<bug-bankaccount-validator-uk at
rt.cpan.org>, or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=BankAccount-Validator-UK>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc BankAccount::Validator::UK

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=BankAccount-Validator-UK>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/BankAccount-Validator-UK>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/BankAccount-Validator-UK>

=item * Search CPAN

L<http://search.cpan.org/dist/BankAccount-Validator-UK/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2017 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License  (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of BankAccount::Validator::UK


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;
use File::ShareDir::Install;

install_share dist => 'share';

WriteMakefile(
    NAME               => 'BankAccount::Validator::UK',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/BankAccount/Validator/UK.pm',
    ABSTRACT_FROM      => 'lib/BankAccount/Validator/UK.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker'     => 0,
        'File::ShareDir::Install' => '0.13',
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'autodie'              => '2.26',
        'Moo'                  => '2.000000',
        'namespace::autoclean' => '0.28',
        'File::ShareDir'       => '1.114',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'BankAccount-Validator-UK-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'BankAccount::Validator::UK'       => { file => 'lib/BankAccount/Validator/UK.pm'     , version => '0.54' },
            'BankAccount::Validator::UK::Rule' => { file => 'lib/BankAccount/Validator/UK/Rule.pm', version => '0.54' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/BankAccount-Validator-UK.git',
                web  => 'https://github.com/manwar/BankAccount-Validator-UK',
            },
        }})
     : ()
    ),
);

package MY;
use File::ShareDir::Install 'postamble';
